.NET SDK

jpcoenen opened the pull request:
This PR is created to allow for reviewing of the current state of the `swig` branch.

Marton6 committed:
Remove previous implementation

hculea committed:
add go.mod and wrapper

Marton6 committed:
Add error handling workaround

hculea committed:
added proper error handling

Marton6 committed:
Add swig code generation

hculea committed:
Merge pull request #1 from secrethub/feature/add-secrethub-wrapper

SecretHub Wrapper

Marton6 committed:
Return error messages through output parameters

Marton6 committed:
Merge pull request #3 from secrethub/feature/improve-error-handling

Return error messages through output parameters

Marton6 committed:
Rename Read to ReadString

Marton6 committed:
Return secret version and value on Read

hculea committed:
implemented the last 4 pending functions

hculea committed:
fixed the 3 nice to have functions.

hculea committed:
revert changes on resolveEnv

hculea committed:
Update output/runme.cs

Co-authored-by: Marton Soos <marton.soos@secrethub.io>

hculea committed:
Merge pull request #4 from secrethub/feature/return-json

`Nice to have` functions

Marton6 committed:
Merge branch 'swig' into feature/read-and-read-string

hculea committed:
resolveEnv returning json string

Marton6 committed:
Add CreatedAt field to Secret

Marton6 committed:
Change CreatedAt type to long long

Marton6 committed:
Run gofmt

Marton6 committed:
Fix indentation

Marton6 committed:
Add SecretVersionID

Marton6 committed:
Add status field

Marton6 committed:
Rename Secret to SecretVersion

Marton6 committed:
Make uuid and time separate C types

Marton6 committed:
Add Secret embedded in SecretVersion type

Marton6 committed:
Add comments

Marton6 committed:
Delete unneded file

Marton6 committed:
Merge pull request #5 from secrethub/feature/read-and-read-string

Create `Read` and `ReadString` functions

edif2008 committed:
Add a Makefile

edif2008 committed:
Add newline

edif2008 committed:
Create environment for creating and publishing nuget package

edif2008 committed:
Add app-info to the wrapper

edif2008 committed:
Adjust Makefile based on review

Marton6 committed:
Implement make nupkg command

edif2008 committed:
Merge pull request #7 from secrethub/feature/makefile

Add Makefile

edif2008 committed:
Merge branch 'feature/makefile' into feature/nuget

edif2008 committed:
Adjust Makefile to make nuget packaging work

edif2008 committed:
Merge branch 'feature/makefile' into feature/app-info

edif2008 committed:
Improve readibility in creating the client

Marton6 committed:
Clean up after nuget package creation

Marton6 committed:
Merge pull request #8 from secrethub/feature/nuget

Build NuGet Package

Marton6 committed:
Merge branch 'swig' into feature/app-info

Marton6 committed:
Add comment

Marton6 committed:
Merge pull request #9 from secrethub/feature/app-info

App info

Marton6 committed:
Change client version from 1.0.0 to 0.1.0

Marton6 committed:
Merge pull request #10 from secrethub/feature/fix-client-version

Change client version from 1.0.0 to 0.1.0

Marton6 committed:
Rename secrethub class to Client and put Client in SecretHub namespace

Marton6 committed:
Build and include .dll in nuget package

Marton6 committed:
Add deps command for installing compilers

Marton6 committed:
Fix dll name: libClient.dll -> Client.dll

edif2008 committed:
Merge pull request #12 from secrethub/feature/build-nuget-for-windows

Build NuGet package for Windows (amd64)

edif2008 committed:
Simplify Makefile

edif2008 committed:
Test suite for .NET

edif2008 committed:
Merge branch 'swig' into feature/add-resolve-env

edif2008 committed:
Add a missing semi colon

edif2008 committed:
Rename dotnet folder variable

edif2008 committed:
Merge pull request #13 from secrethub/feature/simplify-makefile

Simplify Makefile

edif2008 committed:
Adjust tests and makefile based on review

edif2008 committed:
Merge branch 'swig' into feature/test

edif2008 committed:
Adjust csproj file to ignore the unit test files when making the nupkg

edif2008 committed:
Complete deps function

edif2008 committed:
Use the official installation of swig

edif2008 committed:
Match only the code error

edif2008 committed:
Merge branch 'swig' into feature/add-resolve-env

edif2008 committed:
Use Resolve and ResolveEnv from secrethub-go SDK

edif2008 committed:
Enhance Makefile functionality

Marton6 committed:
Merge pull request #14 from secrethub/feature/test

Test suite for .NET

Marton6 committed:
Revert "Use Resolve and ResolveEnv from secrethub-go SDK"

We decided to postpone adding these functions to the secrethub-go SDK.
This reverts commit 99a722f5ad9843357cff184353413dd1bc7862d3.

Marton6 committed:
Free C.Strings in ResolveEnv

Marton6 committed:
Merge branch 'swig' into feature/add-resolve-env

Marton6 committed:
Fix freeing CStrings in ResolveEnv

Marton6 committed:
Decode json to C# Dictionary and add test case

Marton6 committed:
Merge pull request #6 from secrethub/feature/add-resolve-env

ResolveEnv

edif2008 committed:
Rename extracted folder to tmp

jpcoenen commented on dotnet/test/UnitTest.cs:
```diff
@@ -0,0 +1,98 @@
+using Xunit;
+using System.Text.RegularExpressions;
+
+namespace SecretHubTest
```
These tests currently can only be executed by people who have access to the hardcoded repository. I can imagine a few ways to make them runnable by anyone:

1. Make the repository configurable and provide a setup script (or better, make the setup part of the tests).
2. Spin up a real server for the tests (in practice not really desirable because the server code is not public).
3. Introduce a mock server to run these tests against (not trivial to set up though).

Or do you maybe already have something in mind to make this more reproducible?

edif2008 replied:
For now it looks hardcoded since we build and check the library itself instead of the deliverable, which will be the NuGet package. 
The way I can see poeple testing this once the package is available (either locally or online) is by just adding the dependency to the library and installing the package. After that, no hardcoded repo is needed. They can use the test suite in any C# project.
I also have an idea of enabling people to test the package themselves in the repo by changing the script to create the NuGet package and install it automatically. However, I am not sure if that solves the hardcoded repo thing you have in mind, since it may seem similar to what we have now, just that it focuses of testing functionality of the NuGet package, which will be what we intend to release.

Marton6 replied:
I am in favor of setting up the test repo in the test suite. But do you think this needs to be done before the release, @jpcoenen?

jpcoenen replied:
> But do you think this needs to be done before the release, @jpcoenen?

As long as we make sure that it's there for a public release, it should be fine.

jpcoenen commented on secrethub_wrapper.go:
```diff
@@ -0,0 +1,187 @@
+			Version: "0.1.0",
+		}),
+	}
+	client, err := secrethub.NewClient(options...)
+	return client, err
```
```suggestion
	return secrethub.NewClient(options...)
```

edif2008 replied:
Fixed.

jpcoenen commented on secrethub_wrapper.go:
```diff
@@ -0,0 +1,187 @@
+		*errMessage = C.CString(err.Error())
+		return nil
+	}
+	bits := strings.Split(C.GoString(ref), "://")
+	if len(bits) == 2 && bits[0] == "secrethub" {
+		secret, err := client.Secrets().Read(bits[1])
```
Go has some handy helper functions for prefixes :grin: 
```suggestion
	lowercaseRef := strings.ToLower(C.GoString(ref))
	prefix := "secrethub://"
	if strings.HasPrefix(lowercaseRef, prefix) {
		secret, err := client.Secrets().Read(strings.TrimPrefix(lowercaseRef, prefix))
```

_Haven't compiled this, might contain some typos._

edif2008 replied:
Nice thing to see. Adjusted to the suggested implementation.

jpcoenen commented on secrethub_wrapper.go:
```diff
@@ -0,0 +1,187 @@
+	resolvedEnv := make(map[string]string, len(envVars))
+	for _, value := range envVars {
+		envVar := strings.Split(value, "=")
+		key := envVar[0]
+		value := C.CString(envVar[1])
```
Lines in `os.Environ()` are not guaranteed to have an `=` in them. If they don't, this causes a panic.

Can we just ignore all lines where the length of `envVar` is `1`?

edif2008 replied:
According to the documentation, [os.Environ()](https://golang.org/pkg/os/#Environ) returns in `key=value` format. Therefore, there should be no concerns about lines without an `=` in them.
However, in the adjustments that I will make, we will ignore all lines where the length ov `envVar` is 1, even though this will never happen.

jpcoenen replied:
Hmm, you're right. 

I was confused by the weird entry that Windows sometimes adds to the `Environ()` list and that has previously bitten up (because it's really weird): `:=::=`. But entry still adheres to the rule that every entry contains a `=`. 

Nevertheless, the `== 1` check does no harm, so we can still leave it in.

edif2008 replied:
The condition is there.

jpcoenen commented on secrethub_wrapper.go:
```diff
@@ -0,0 +1,187 @@
+	envVars := os.Environ()
+	resolvedEnv := make(map[string]string, len(envVars))
+	for _, value := range envVars {
+		envVar := strings.Split(value, "=")
```
What happens if the value of an environment variable contains a `=`? I think we want to use `strings.SplitN(value, "=", 2)` here.

edif2008 replied:
Interesting idea. Fixed.

mackenbach replied:
@edif2008 consider adding a test for these cases.

edif2008 replied:
Added extra case for more equals here: #23

jpcoenen commented on secrethub_wrapper.go:
```diff
@@ -0,0 +1,187 @@
+	}
+	err = client.Secrets().Delete(C.GoString(path))
+	if err != nil {
+		*errMessage = C.CString(err.Error())
```
Though technically not necessary, to avoid future bugs, it's a good practice to include a `return` here. If I were to something after this line, I would silently introduce a bug, because the code would get executed, even when `Delete()` returns an error:

```golang
if err != nil {
		*errMessage = C.CString(err.Error())
}
fmt.Println("All went well!") // Well yes, but actually no.
```
And best thing of all: it won't hurt the performance. The compiler will almost certainly not make a difference between there being a `return` or not.

```suggestion
		*errMessage = C.CString(err.Error())
		return
```

edif2008 replied:
Nice catch. Fixed.

jpcoenen commented on secrethub_wrapper.go:
```diff
@@ -0,0 +1,187 @@
+	}
+	_, err = client.Secrets().Write(C.GoString(path), []byte(C.GoString(secret)))
+	if err != nil {
+		*errMessage = C.CString(err.Error())
```
As explained above:
```suggestion
		*errMessage = C.CString(err.Error())
		return
```

edif2008 replied:
Fixed.

jpcoenen commented on secrethub_wrapper.go:
```diff
@@ -0,0 +1,187 @@
+)
+
+// Client creates and returns a new Go client to be used by xgo clients.
+func Client() (*secrethub.Client, error) {
```
This causes us to continuously create a new client. This does not necessarily have to be an issue, but for example in the case of using any of the Cloud Identity Providers, this means that the authentication is triggered everytime `Client()` is called. Especially in the way `Client()` is currently called in `ResolveEnv` this can be problematic.

Is there any way in which we can introduce some state to resolve this? In the Go SDK we tackle this by requiring the user to create an instance of the client with `secrethub.NewClient()`. I can imagine that is a little less trivial using swig.

SimonBarendse replied:
Another advantage of re-using the client: we can add caching to the (Golang) SDK to optimize repeating (similar) operations.

Marton6 replied:
This was on the list of future improvements, but it is indeed quite problematic in `ResolveEnv`.

I found a way to allow the user to create a new client object and then call these methods on it (and also clean up the client afterwards). However this will take significant effort to implement (~6h maybe). Do we want this for the alpha release? Or would "fixing" the `ResolveEnv` function (making it use one client) also suffice?

Marton6 replied:
I implemented the previously mentioned change. The functions are now methods on the `client` object, the underlying go clients are cached and a go client is only created when in the C# code the client constructor is called.

jpcoenen commented on secrethub_wrapper.go:
```diff
@@ -0,0 +1,187 @@
+			Version: "0.1.0",
+		}),
+	}
+	client, err := secrethub.NewClient(options...)
```
What if the user wants to set any other options? Can we imagine any other way to set them besides environment variables? This is probably very much related to the above question about continuously recreating the client.

Marton6 replied:
As mentioned in my other comment, I found a way to make this possible but it will take significant effort (~6h). Do you think we want to have this before the alpha release?

Marton6 replied:
I implemented the required changes to make the code forward compatible with adding options.

jpcoenen commented on secrethub_wrapper.go:
```diff
@@ -0,0 +1,187 @@
+// has the format `secrethub://<path>`. Otherwise it returns `ref` unchanged, as an array of bytes.
+//export Resolve
+func Resolve(ref *C.char, errMessage **C.char) *C.char {
+	client, err := Client()
+	if err != nil {
+		*errMessage = C.CString(err.Error())
+		return nil
+	}
```
Do we have to call this here or can we also move this to the point where we know we're dealing with a reference to a secret?

edif2008 replied:
Indeed, we dont. Moved it in the right place.

jpcoenen commented on Makefile:
```diff
@@ -0,0 +1,65 @@
+
+.PHONY: deps
+deps:
+	sudo apt install gcc
+	sudo apt install gcc-mingw-w64
```
This can be solved by using Docker, but `apt` only works on Debian-like systems. For example, my Arch machine does not have `apt`, making it difficult to reproduce your work.

Marton6 replied:
Dockerization of the build process is planned, but I'd say we leave it for after the alpha release.

jpcoenen reviewed (commented):
Nice work @hculea @edif2008 @Marton6!

Had a look at the current `swig` branch and wrote down some of my thoughts. Some points you might have already thought about or may even have already discussed through some other channel, but I thought it would be good to record them all in a single place and have room for a discussion.

jpcoenen commented on dotnet/secrethub.i:
```diff
@@ -0,0 +1,82 @@
+%module Client
```
An open question is whether we want to call this `Client` or something like `SimpleClient` or `LiteClient` to indicate that it does not have to full functionality of the SecretHub client.

@florisvdg @SimonBarendse any thoughts on this?

SimonBarendse replied:
What does the migration path look like **if** we want to add more functionality? Can we easily keep the existing functionality available at `LiteClient`, while we add more functionality in `Client`?

I'm not sure if it's necessary to indicate the incompleteness in the import path itself. I've seen many SDKs indicating supported endpoints/operations in the readme of the repo. We could adopt a similar approach.

jpcoenen replied:
One of the consideration that I did not make clear in my message is the way this client is initialized: it can only be initialized with default options and is not a full-fledged client in that way. But that's also something that we can try to fix (see https://github.com/secrethub/secrethub-xgo/pull/16#discussion_r488655898)

florisvdg replied:
But is it then even a client if it's all managed and created for you under the hood anyway? Shouldn't it just be `SecretHub.Read` in that case?

Because if we bring the fact that it's a `Client` to the forefront, you may expect more flexibility which is not there at the moment (and out of scope for the initial release imo).

florisvdg replied:
And when we do add a more powerful client, it will still be nice to have a basic `SecretHub.Read`, so we won't even have to deprecate that.

jpcoenen replied:
Maybe that's the better approach here. Though I am not sure whether that is technically feasible. @Marton6 probably knows that best.

SimonBarendse replied:
I like that idea @florisvdg

Marton6 replied:
I changed the code generation such that now all functions are methods on the `Client` object:
```c#
var client = new SecretHub.Client();
string secret = client.ReadString("secrethub-xgo/dotnet/test-secret");
```

This is forward compatible with adding client options and more methods, so I'd say it's safe to call it `Client` (the module name is different now but it does not matter). So I suggest we have a `SecretHub` namespace with a `Client` class in it.

SimonBarendse replied:
That sounds good to me

edif2008 committed:
Make first adjustments based on Joris' feedback

edif2008 commented:
Thank you so much for the feedback, @jpcoenen.
The necessary adjustments based on your feedback are in this PR: #18 
For now, most of the small thing have been fixed. Other things that need a bit of thinking (e.g dockerfile, client options, client creation repetition) will be added there as soon as possible

edif2008 committed:
Fix build failure

jpcoenen committed:
Merge pull request #18 from secrethub/feature/major-adjustments

edif2008 committed:
Fix nupkg build error

edif2008 committed:
Make first prototype of documentation

edif2008 committed:
Add Json.Net dependency in the documentation

Marton6 committed:
Merge pull request #15 from secrethub/feature/deps

Complete deps functionality

Marton6 committed:
Merge pull request #19 from secrethub/feature/fix-nupkg

Fix nupkg build error

Marton6 committed:
Change NuGet Package version to 0.1.0

edif2008 committed:
Merge pull request #21 from secrethub/fix/version

Change NuGet Package version to 0.1.0

edif2008 committed:
Merge branch 'swig' into feature/documentation

edif2008 committed:
Merge branch 'swig' into feature/enhance-makefile

edif2008 committed:
Add dotnet dependency

edif2008 committed:
Small adjustments

edif2008 committed:
Add extra test case with more equal signs

Marton6 committed:
Add client constructor and destructor

This commit adds constructor and destructor functions for the SecretHub
client and configures them to be called from C# when a new C# client is
created or when one is destroyed.

I tested the destruction of the client object and resources are being
freed correctly.

SimonBarendse committed:
Merge pull request #23 from secrethub/feature/tests

Add extra test case with more equal signs

Marton6 committed:
Rename module to SecretHubXGO and MyClient to Client

Marton6 committed:
Make all functions methods on the client

This PR makes all functions methods on the client and updates
the test cases to reflect this change

Marton6 committed:
Merge branch 'swig' into feature/configurable-client

Marton6 committed:
Hide ID field of Client

Marton6 committed:
Merge pull request #24 from secrethub/feature/configurable-client

Client class

Marton6 commented on secrethub_wrapper.go:
```diff
@@ -0,0 +1,196 @@
+)
+
+var clients = make(map[int]secrethub.ClientInterface)
+var nextClientID = 1;
```
This is an int so there is technically a limit of `4294966654` clients you can have during the runtime of the program (if you do not delete the first one). I assume this is not a big issue for now, but let me know if it should be fixed.

Marton6 commented on secrethub_wrapper.go:
```diff
@@ -0,0 +1,196 @@
+func GetGoClient(cClient *C.struct_Client) secrethub.ClientInterface {
+	return clients[int(cClient.ID)]
+}
+
+
```
```suggestion

```

Marton6 committed:
Add comments and rename method

Marton6 committed:
Merge pull request #25 from secrethub/feature/add-comments

Add comments

Marton6 committed:
Synchronize creation of clients

edif2008 committed:
Wrap up a good working enhanced makefile

edif2008 committed:
Merge branch 'swig' into feature/enhance-makefile

edif2008 committed:
Merge branch 'swig' into feature/documentation

edif2008 committed:
Adjust guide based on review

edif2008 committed:
Put the content of the guide back into the readme file

edif2008 committed:
Remove the step of installing Json.net dependency

Marton6 committed:
Merge pull request #26 from secrethub/feature/concurrency-safe-constructor

Concurrency-safe constructor

Marton6 committed:
Use sync.Map to store clients

Marton6 committed:
Fix error not returned from resolveEnv

Marton6 committed:
Use 64 bit client ids

Marton6 committed:
Merge pull request #28 from secrethub/feature/concurrent-map

Concurrent Client map

Marton6 committed:
Merge pull request #29 from secrethub/fix/resolve-env

Fix ResolveEnv

Marton6 committed:
Merge pull request #30 from secrethub/fix/long-client-id

Use 64bit client ids

Marton6 committed:
Target netstandard2.0

Marton6 committed:
Uncomment nupkg-publish make command

Marton6 committed:
Update package description

Marton6 committed:
Update description

edif2008 committed:
Adjust readme based on Joris' review

edif2008 committed:
Add table of contents to readme

edif2008 committed:
Merge branch 'swig' into feature/documentation

edif2008 committed:
Adjust exception and rersolveEnv decumentation

edif2008 committed:
Adjust based on Marton's feedback

edif2008 committed:
Fix punctuation

jpcoenen committed:
Minor changes to readme

As suggested in the PR.

jpcoenen committed:
Move the Golang dependency to build from source

jpcoenen committed:
Move the Golang dependency to build from source

jpcoenen committed:
Merge pull request #20 from secrethub/feature/documentation

jpcoenen committed:
Merge pull request #31 from secrethub/feature/alpha-release

Marton6 committed:
Merge branch 'swig' into feature/enhance-makefile

edif2008 committed:
Merge pull request #22 from secrethub/feature/deps-dotnet

Add dotnet dependency

edif2008 committed:
Merge branch 'swig' into feature/enhance-makefile

edif2008 committed:
Merge pull request #27 from secrethub/feature/enhance-makefile

Enhance makefile

jpcoenen commented:
This was meant as a way to review the project. Closing it as the work is merged through other PR's.
